src/ucp/core/ucp_mm.c

```diff
@@ -1201,27 +1180,38 @@ ucs_status_t ucp_mem_type_reg_buffers(ucp_worker_h worker, void *remote_addr,
     char *rkey_buffer;
 
     if (!(md_attr->flags & UCT_MD_FLAG_NEED_RKEY)) {
-        rkey_bundle->handle = NULL;
-        rkey_bundle->rkey   = UCT_INVALID_RKEY;
-        status              = UCS_OK;
+        pack_context->rkey_bundle.handle = NULL;
+        pack_context->rkey_bundle.rkey   = UCT_INVALID_RKEY;
+        status                           = UCS_OK;
         goto out;
     }
 
-    memh->super.super.start = (uintptr_t)remote_addr;
-    memh->super.super.end   = (uintptr_t)remote_addr + length;
-    ucp_memh_init(memh, context, 0, UCT_MD_MEM_ACCESS_ALL,
-                  UCT_ALLOC_METHOD_LAST, mem_type);
-
     tl_md  = &context->tl_mds[md_index];
     cmpt   = context->tl_cmpts[tl_md->cmpt_index].cmpt;
-    status = ucp_memh_init_uct_reg(context, memh, UCS_BIT(md_index),
-                                   UCT_MD_MEM_ACCESS_ALL, "mem_type");
-    if (status != UCS_OK) {
-        goto out;
+    if (((context->cache_md_map[mem_type] & UCS_BIT(md_index)) == 0) ||
+        (context->rcache == NULL)) {
```

Comment:
maybe we could set cache_md_map to 0 for all, when rcache==NULL, to avoid this check?